src/walk.rs

```diff
@@ -365,188 +347,230 @@ fn spawn_receiver(
-            }
 
-            // Filter out unwanted modification times
-            if !config.time_constraints.is_empty() {
-                let mut matched = false;
-                if let Some(metadata) = entry.metadata() {
-                    if let Ok(modified) = metadata.modified() {
-                        matched = config
-                            .time_constraints
-                            .iter()
-                            .all(|tf| tf.applies_to(&modified));
+                if config.is_printing() {
+                    if let Some(ls_colors) = &config.ls_colors {
+                        // Compute colors in parallel
+                        entry.style(ls_colors);
                     }
                 }
-                if !matched {
-                    return ignore::WalkState::Continue;
+
+                let send_result = tx.send(WorkerResult::Entry(entry));
+
+                if send_result.is_err() {
+                    return WalkState::Quit;
                 }
-            }
 
-            if config.is_printing() {
-                if let Some(ls_colors) = &config.ls_colors {
-                    // Compute colors in parallel
-                    entry.style(ls_colors);
+                // Apply pruning.
+                if config.prune {
+                    return WalkState::Skip;
                 }
-            }
 
-            let send_result = tx_thread.send(WorkerResult::Entry(entry));
+                WalkState::Continue
+            })
+        });
+    }
 
-            if send_result.is_err() {
-                return ignore::WalkState::Quit;
-            }
+    /// Perform the recursive scan.
+    fn scan(&self, paths: &[PathBuf]) -> Result<ExitCode> {
+        let config = &self.config;
+        let walker = self.build_walker(paths)?;
 
-            // Apply pruning.
-            if config.prune {
-                return ignore::WalkState::Skip;
-            }
+        if config.ls_colors.is_some() && config.is_printing() {
+            let quit_flag = Arc::clone(&self.quit_flag);
+            let interrupt_flag = Arc::clone(&self.interrupt_flag);
+
+            ctrlc::set_handler(move || {
+                quit_flag.store(true, Ordering::Relaxed);
 
-            ignore::WalkState::Continue
-        })
-    });
+                if interrupt_flag.fetch_or(true, Ordering::Relaxed) {
+                    // Ctrl-C has been pressed twice, exit NOW
+                    ExitCode::KilledBySigint.exit();
+                }
+            })
+            .unwrap();
+        }
+
+        // Channel capacity was chosen empircally to perform similarly to an unbounded channel
+        let (tx, rx) = bounded(0x4000 * config.threads);
+
+        let exit_code = thread::scope(|scope| {
+            // Spawn the receiver thread(s)
+            let receiver = scope.spawn(|| self.receive(rx));
+
+            // Spawn the sender threads.
+            self.spawn_senders(walker, tx);
+
+            receiver.join().unwrap()
```

Comment:
instead of spawning a separate thread and joining on it here, could we just call `self.receive(rx)` here?

Reply:
I tried that originally but the problem is `spawn_senders()` is blocking.  So we need an extra background thread either for `receive()` or `spawn_senders()`.